<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Testing generators with promises</title>
  </head>
  <body>
    <script>
      function getJSON(url) {
        return new Promise((resolve, reject) => {
          const request = new XMLHttpRequest();

          // Initialize the request
          request.open('GET', url);

          // Register onload handler that will be called if the server has responded
          request.onload = function () {
            try {
              if (this.status === 200) {
                resolve(JSON.parse(this.response));
              } else {
                reject(this.status + ' ' + this.statusText);
              }
            } catch (e) {
              reject(e.message);
            }
          };

          // If there's an error while communicating with the server, reject
          request.onerror = function () {
            reject(this.status + ' ' + this.statusText);
          };

          // Send the request
          request.send();
        });
      }

      asynchronous(function* () {
        console.log('Making generator');
        try {
          const posts = yield getJSON(
            'https://jsonplaceholder.typicode.com/posts'
          );
          console.log(posts);
          const post = yield getJSON(
            'https://jsonplaceholder.typicode.com/posts/1'
          );
          console.log(post);
        } catch (e) {
          // Oh no, something wrong with the json
          console.log(e);
        }
      });

      // Helper function that will control our generator
      function asynchronous(generator) {
        // Creates an iterator through which we'll control the generator
        // This iterator will be used to resume iterator execution
        var iterator = generator();

        // Defines a function that will handle
        // one returned value by a generator
        function handle(iteratorResult) {
          // Stops when the generator has no more results
          if (iteratorResult.done) {
            console.log(iteratorResult);
            return;
          }

          const iteratorValue = iteratorResult.value;

          if (iteratorValue instanceof Promise) {
            iteratorValue
              .then(res => handle(iterator.next(res)))
              .catch(err => iterator.throw(err));
          }
        }
        try {
          handle(iterator.next());
        } catch (e) {
          iterator.throw(e);
        }
      }
    </script>
  </body>
</html>
